Runtime/C++/src/bebop.hpp

```diff
@@ -0,0 +1,292 @@
+        m_pointer += 2;
+        return v;
+#else
+        const uint16_t b0 = *m_pointer++;
+        const uint16_t b1 = *m_pointer++;
+        return (b1 << 8) | b0;
+#endif
+    }
+
+    uint32_t readUint32() {
+#if BEBOP_ASSUME_LITTLE_ENDIAN
+        const auto v = *reinterpret_cast<const uint32_t*>(m_pointer);
+        m_pointer += 4;
+        return v;
+#else
+        const uint32_t b0 = *m_pointer++;
+        const uint32_t b1 = *m_pointer++;
+        const uint32_t b2 = *m_pointer++;
+        const uint32_t b3 = *m_pointer++;
+        return (b3 << 24) | (b2 << 16) | (b1 << 8) | b0;
+#endif
+    }
+
+    uint64_t readUint64() {
+#if BEBOP_ASSUME_LITTLE_ENDIAN
+        const auto v = *reinterpret_cast<const uint64_t*>(m_pointer);
+        m_pointer += 8;
+        return v;
+#else
+        const uint64_t b0 = *m_pointer++;
+        const uint64_t b1 = *m_pointer++;
+        const uint64_t b2 = *m_pointer++;
+        const uint64_t b3 = *m_pointer++;
+        const uint64_t b4 = *m_pointer++;
+        const uint64_t b5 = *m_pointer++;
+        const uint64_t b6 = *m_pointer++;
+        const uint64_t b7 = *m_pointer++;
+        return (b7 << 0x38) | (b6 << 0x30) | (b5 << 0x28) | (b4 << 0x20) | (b3 << 0x18) | (b2 << 0x10) | (b1 << 0x08) | b0;
+#endif
+    }
+
+    int16_t readInt16() { return static_cast<uint16_t>(readUint16()); }
+    int32_t readInt32() { return static_cast<uint32_t>(readUint32()); }
+    int64_t readInt64() { return static_cast<uint64_t>(readUint64()); }
+
+    float readFloat32() {
+        const auto v = readUint32();
+        return reinterpret_cast<const float&>(v);
+    }
+
+    double readFloat64() {
+        const auto v = readUint64();
+        return reinterpret_cast<const double&>(v);
+    }
+
+    bool readBool() {
+        return readByte() != 0;
+    }
+
+    std::vector<uint8_t> readBytes() {
+        const auto length = readUint32();
+        std::vector<uint8_t> v(m_pointer, m_pointer + length);
+        m_pointer += length;
+        return v;
+    }
+
+    std::string readString() {
+        const auto length = readUint32();
+        std::string v(m_pointer, m_pointer + length);
+        m_pointer += length;
+        return v;
+    }
+
+    Guid readGuid() {
+        Guid guid { m_pointer };
+        m_pointer += 16;
+        return guid;
+    }
+
+    // Read a date (as ticks since the Unix Epoch).
+    TickDuration readDate() {
+        const uint64_t ticks = readUint64() & 0x3fffffffffffffff;
+        return TickDuration(ticks - ticksBetweenEpochs);
+    }
+
+    uint32_t readMessageLength() { return readUint32(); }
+};
+
+class BebopWriter {
+    std::unique_ptr<std::vector<uint8_t>> m_buffer;
+    BebopWriter() {}
+public:
+    BebopWriter(BebopWriter const&) = delete;
+    void operator=(BebopWriter const&) = delete;
+    static BebopWriter& instance() {
+        static BebopWriter instance;
+        instance.m_buffer = std::make_unique<std::vector<uint8_t>>();
```

Comment:
Does it make sense to do something like:
```
instance.m_buffer.reserve(1024);
```
Just to avoid having to reallocate and copy repeatedly when doing the push_back() calls below.

Perhaps if you make this class not a Singleton anymore, then a parameter to the constructor could be an initial reserve size with a default value of something if it is not provided.